driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/BatchCursorFlux.java

```diff
@@ -94,6 +94,8 @@ private void recurseCursor(){
                         })
                         .subscribe();
                 }
+        } else if (sink.isCancelled()) {
+            closeCursor();
```

Comment:
I am confused. My understanding previously was that we register `sink.onDispose(this::closeCursor)` specifically to call `closeCursor` when `sink` is either terminated with `onError`, or with `onComplete`, or a a result of being cancelled. This means:
1. We do not need to sprinkle explicit `closeCursor` in our code. Doing so confuses reader and makes it harder to reason about the code, because it creates expectations that `sink.onDispose(this::closeCursor)` is not enough.
2. We do not need `sink.onCancel(this::closeCursor)`.

We would have a good reason to write something beyond `sink.onDispose(this::closeCursor)` (like registering `sink.onCancel`, or calling `closeCursor` explicitly) if we knew about a bug in Project Reactor and was trying to work our way around it. Is this the case?

Reply:
I think this fixes a possible race consider:

```
Subscription.request(1)
Subscription.cancel()
```

```
------- batchCursorPublisher.subscribe called
--------------- Close cursor called (batchCursor == null)
----------------------------- batchCursorPublisher.subscribe completes - recurseCursor closes the cursor.
```

Reply:
I see, you are right! 👏  There is a race condition (but not a data race), that allows to have `batchCursor` left open despite `sink` being cancelled. I spent some significant time trying to prove that the solution you propose is correct, and I failed. Instead, I proved that there are allowed executions in which we will leave `batchCursor` open.

----
### Labels:
- `so` - synchronization order.
- `po` - program order.
- `hb` - happens-before order.
- `w(batchCursor)` - the write action that initializes `batchCursor` as a result of calling `subscribe`. 
- `r_in_closeCursor(batchCursor)` - the read from `batchCursor` action in `closeCursor` that is used to check for null.
- `w(sink.cancelled)` - a synchronization action (equivalent to volatile write) somewhere in the `sink` code that results in `sink.isClosed` returning true. This actions is done as a result of cancelling `sink`.
- `r(sink.cancelled)` - a synchronization action (equivalent to volatile read) somewhere in the `sink.isClosed` code that determines the return value of this method. This action is done as a result of the `sink.isCancelled()` call we do right after (in `po`) the action `w(batchCursor)`.

### Facts:
1. `hb`: `w(sink.cancelled)`, `r_in_closeCursor(batchCursor)` because `sink` calls `closeCursor` as a result of being cancelled.
2. `po`: `w(batchCursor)`, `r(sink.cancelled)` trivially from how the code is written.
3. `hb`: `w(batchCursor)`, `r(sink.cancelled)` because of fact 2.
5. `w(batchCursor)`, `r_in_closeCursor(batchCursor)`, `w(sink.cancelled)`, `r(sink.cancelled)` are synchronization actions, which means they are totally ordered in `so`. Executions with any `so` that does not result in violating JMM constraints is allowed.
6. `so`: `w(batchCursor)`, `r(sink.cancelled)` because of facts 5 and 2 (`so` is consistent with `po`).

### Proof:
Because of the fact 4, all allowed executions can be classified into the following two groups:
1. `so`: `w(batchCursor)`, `r_in_closeCursor(batchCursor)`
2. `so`: `r_in_closeCursor(batchCursor)`, `w(batchCursor)`

#### Executions in group 1
`hb`: `w(batchCursor)`, `r_in_closeCursor(batchCursor)` because of the `so` in this group of executions.
The code in `closeCursor` observes non-null `batchCursor` and calls `batchCursor.close` in all executions in this group.

#### Executions in group 2
```
hb: w(sink.cancelled), r_in_closeCursor(batchCursor)                                    fact 1
so:                    r_in_closeCursor(batchCursor), w(batchCursor)                    for this group of executions
hb:                                                   w(batchCursor), r(sink.cancelled) fact 3
so:                                                   w(batchCursor), r(sink.cancelled) fact 6
```
Or, with arrows:
```
w(sink.cancelled) -hb-> r_in_closeCursor(batchCursor) -\
                                                       so
                                                         \-> w(batchCursor) -po, hb, so-> r(sink.cancelled)
```
Unfortunately, there is nothing in the orders specified above that would either give us `hb`: `w(sink.cancelled)`, `r(sink.cancelled)`, or prohibit `so`: `r(sink.cancelled)`, `w(sink.cancelled)`.

As a result, executions in which `r_in_closeCursor(batchCursor)` does not observe `w(batchCursor)`, and `r(sink.cancelled)` does not observe `w(sink.cancelled)` are allowed (they constitute a subset of executions in group 2). Such executions leave `batchCursor` not closed despite `sink` being cancelled.

Reply:
We can prevent faulty executions described above by enclosing `r_in_closeCursor(batchCursor)` and `w(batchCursor) & r(sink.cancelled)` in `synchronized` blocks (or equivalent constructs). Once we do that, the following will be true for all the previously faulty executions: from `so: r_in_closeCursor(batchCursor), w(batchCursor)` follows `hb: r_in_closeCursor(batchCursor), w(batchCursor)` (this is not the case with the current code). As a result, we will start having `hb`: `w(sink.cancelled)`, `r(sink.cancelled)` for all the previously faulty executions, and they will stop being faulty.

Reply:
I gave this situation more thought, and it seems like the happens-before order is provable consistent with the synchronization order: [is HB consistent with SO for not otherwise related synchronization actions?](http://cs.oswego.edu/pipermail/concurrency-interest/2021-February/017286.html), [proof](http://cs.oswego.edu/pipermail/concurrency-interest/2021-February/017287.html) (hopefully, a correct one).

Given this new proved fact, we can come back to the reasoning about executions in group 2.

#### Executions in group 2
```
w(sink.cancelled) -hb-> r_in_closeCursor(batchCursor) -\
                                                       so
                                                         \-> w(batchCursor) -po, hb, so-> r(sink.cancelled)
```
gives us
```
w(sink.cancelled) -so, hb-> r_in_closeCursor(batchCursor) -\
                                                       so
                                                         \-> w(batchCursor) -po, hb, so-> r(sink.cancelled)
```
The above gives `so`: `w(sink.cancelled)`, `r(sink.cancelled)` and, therefore `hb`: `w(sink.cancelled)`, `r(sink.cancelled)`. The code initializing `batchCursor` observes that `sink` is cancelled and calls `batchCursor.close` in all executions in this group.

Reply:
@rozza Sorry for lots of text, but now we have a formal proof that the current approach of eliminating that race condition you mentioned is correct.

The only thing I'd like to propose, is to remove the `if (sink.isCancelled()) closeCursor()` code from `recurseCursor` and put it right after `batchCursor = bc`. This way we do not introduce unnecessary code to the `recurseCursor`. It may also be useful to explain why `if (sink.isCancelled()) closeCursor()` exists in the code, because this race condition is not so easy to notice.

Reply:
Done.

Reply:
Thank you!